packages/fetchai/connections/prometheus/connection.py

```diff
@@ -0,0 +1,314 @@
+    """The dialogues class keeps track of all prometheus dialogues."""
+
+    def __init__(self, **kwargs) -> None:
+        """
+        Initialize dialogues.
+
+        :return: None
+        """
+
+        def role_from_first_message(  # pylint: disable=unused-argument
+            message: Message, receiver_address: Address
+        ) -> BaseDialogue.Role:
+            """Infer the role of the agent from an incoming/outgoing first message
+
+            :param message: an incoming/outgoing first message
+            :param receiver_address: the address of the receiving agent
+            :return: The role of the agent
+            """
+            # The server connection maintains the dialogue on behalf of the agent
+            return PrometheusDialogue.Role.SERVER
+
+        BasePrometheusDialogues.__init__(
+            self,
+            self_address=str(PUBLIC_ID),
+            role_from_first_message=role_from_first_message,
+            **kwargs,
+        )
+
+
+class PrometheusChannel:
+    """A wrapper for interacting with a prometheus server."""
+
+    THREAD_POOL_SIZE = 3
+
+    def __init__(self, address: Address, metrics: Dict[str, Any], port: int):
+        """
+        Initialize a prometheus channel.
+
+        :param address: The address of the connection.
+        :param metrics: The prometheus metrics.
+        :param port: The port at which to expose the metrics.
+        """
+        self.address = address
+        self.metrics = metrics
+        self._loop: Optional[asyncio.AbstractEventLoop] = None
+        self._queue: Optional[asyncio.Queue] = None
+        self._threaded_pool: ThreadPoolExecutor = ThreadPoolExecutor(
+            self.THREAD_POOL_SIZE
+        )
+        self.logger: Union[logging.Logger, logging.LoggerAdapter] = _default_logger
+        self._dialogues = PrometheusDialogues()
+        self._port = port
+
+    def _get_message_and_dialogue(
+        self, envelope: Envelope
+    ) -> Tuple[PrometheusMessage, Optional[PrometheusDialogue]]:
+        """
+        Get a message copy and dialogue related to this message.
+
+        :param envelope: incoming envelope
+
+        :return: Tuple[Message, Optional[Dialogue]]
+        """
+        message = cast(PrometheusMessage, envelope.message)
+        dialogue = cast(PrometheusDialogue, self._dialogues.update(message))
+        return message, dialogue
+
+    @property
+    def queue(self) -> asyncio.Queue:
+        """Check queue is set and return queue."""
+        if self._queue is None:  # pragma: nocover
+            raise ValueError("Channel is not connected")
+        return self._queue
+
+    async def connect(self) -> None:
+        """
+        Start prometheus http server.
+
+        :return: None
+        """
+        if self._queue:  # pragma: nocover
+            return None
+        self._loop = asyncio.get_event_loop()
+        self._queue = asyncio.Queue()
+        prometheus_client.start_http_server(self._port)
```

Comment:
I confused with this approach developers of this library provided.
you can start it, but no method to stop it, I just checked the code.
anyway the server will shutdown on main app exit, but it can be a problem for tests or several instances of the connection.


probably we can check some wrappers like https://github.com/claws/aioprometheus

Reply:
Yes, exactly. I was also curious why I could not find any `stop_http_server()` function. Thanks for the suggestion!